src/Promitor.Integrations.AzureMonitor/HttpPipelinePolicies/RecordArmRateLimitMetricsPolicy.cs

```diff
@@ -0,0 +1,92 @@
+        private readonly MetricSinkWriter _metricSinkWriter;
+        private readonly IAzureScrapingSystemMetricsPublisher _azureScrapingSystemMetricsPublisher;
+        private const string ThrottlingHeaderName = "x-ms-ratelimit-remaining-subscription-reads";
+        private const string AvailableCallsMetricDescription = "Indication how many calls are still available before Azure Resource Manager (ARM) is going to throttle us.";
+        private const string ThrottledMetricDescription = "Indication concerning Azure Resource Manager are being throttled. (1 = yes, 0 = no).";
+
+        public RecordArmRateLimitMetricsPolicy(string tenantId, string subscriptionId, AzureAuthenticationInfo azureAuthenticationInfo, MetricSinkWriter metricSinkWriter, IAzureScrapingSystemMetricsPublisher azureScrapingSystemMetricsPublisher)
+        {
+            Guard.NotNullOrWhitespace(tenantId, nameof(tenantId));
+            Guard.NotNullOrWhitespace(subscriptionId, nameof(subscriptionId));
+            Guard.NotNull(metricSinkWriter, nameof(metricSinkWriter));
+            Guard.NotNull(azureScrapingSystemMetricsPublisher, nameof(azureScrapingSystemMetricsPublisher));
+            Guard.NotNull(azureAuthenticationInfo, nameof(azureAuthenticationInfo));
+
+            _metricSinkWriter = metricSinkWriter;
+            _azureScrapingSystemMetricsPublisher = azureScrapingSystemMetricsPublisher;
+
+            var id = DetermineApplicationId(azureAuthenticationInfo);
+  
+            _metricLabels = new Dictionary<string, string>
+            {
+                {"tenant_id", tenantId},
+                {"subscription_id", subscriptionId},
+                {"app_id", id},
+            };
+        }
+
+        public override async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
+        {   
+            await ProcessNextAsync(message, pipeline);
+            // Source: https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-manager-request-limits
+            if (message.Response.Headers.Contains(ThrottlingHeaderName))
+            {   
+                message.Response.Headers.TryGetValue(ThrottlingHeaderName, out string remainingApiCallsStr);
+                var subscriptionReadLimit = Convert.ToInt16(remainingApiCallsStr);
+                
+                // Report metric
+                await _metricSinkWriter.ReportMetricAsync(RuntimeMetricNames.RateLimitingForArm, AvailableCallsMetricDescription, subscriptionReadLimit, _metricLabels);
+                await _azureScrapingSystemMetricsPublisher.WriteGaugeMeasurementAsync(RuntimeMetricNames.RateLimitingForArm, AvailableCallsMetricDescription, subscriptionReadLimit, _metricLabels);
+            }
+
+            // throttled by ARM
+            await _azureScrapingSystemMetricsPublisher.WriteGaugeMeasurementAsync(RuntimeMetricNames.ArmThrottled, ThrottledMetricDescription, message.Response.Status == 429 ? 1 : 0, _metricLabels, includeTimestamp: true);
```

Comment:
Move to method and re-use?

Reply:
On second thought, I don't think the synchronous path should be used at all. I decided to throw an exception there instead(and there's no more repeated code ofc), Let me know if you agree.